Copter: Safe RTL

rmackay9 opened the pull request:
This replaces @squilter's earlier PR (https://github.com/ArduPilot/ardupilot/pull/6682) which adds support for "Safe RTL":

- new AC_SafeRTL library does the following:
  - creates and maintains an array of Vector3f objects which are used to store the vehicle's position offset from the EKF origin.
  - SRTL_ACCURACY parameter holds the minimum distance between points
  - SRTL_PATH_LEN holds the maximum number of points recorded (defaults to 200, maximum is 500)
  - update function is called from the main vehicle class at 3.3hz and records the vehicle's position at the end of the array (if necessary)
  - there are a 2 routines run in the IO thread to identify how the path can be simplified:
      - "detect_simplifications" runs in the IO thread and looks for points in a straight line that can thus be removed.  Points that can be removed are recorded in the _simplification_bitmask and _simplification_stack.
     - "detect_loops" looks for loops in the path which are stored in the _prunable_loops array
  - "routine_cleanup" is run regular as part of "update" (at 3hz) to remove points
  - when the SafeRTL flight mode is initiated "through_cleanup" is run as well to ensure all possible simplifications are completed.
- a new SafeRTL flight mode has been added which, when initiated, stops the vehicle then returns the vehicle to the take-off location by using the lib's "pop_point" method to retrieve points from the path.
- the vehicle eventually stops 2 meters above the take-off location and uses regular RTL's Land or Descend feature.

Items potential issues that we might need to address in follow-up PRs:

- at 200 points, SafeRTL consumes about 4k of memory (2.4k for the path itself, 0.8k for each of the simplification-stack and prunable-loops arrays).  Increasing the number of points consumes more memory but also more CPU for the simplification.  We need to test more to find what number of points works best in practice.  Same goes for the ACCURACY parameter.
- the EKF must be reporting a good position at take-off for SafeRTL to be available
- SafeRTL becomes unusable for that flight if the EKF's position estimate goes bad for >15seconds.
- SafeRTL should be made available as a failsafe option.

Some videos:

- sitl test: https://www.youtube.com/watch?v=7YsTC6vd2jE
- actual flight test: https://www.youtube.com/watch?v=Ju0tL-nmQpo

peterbarker commented on libraries/AC_SafeRTL/AC_SafeRTL.h:
`midpoint`?

rmackay9 commented on libraries/DataFlash/LogFile.cpp:
```diff
@@ -1967,3 +1967,19 @@ void DataFlash_Class::Log_Write_Proximity(AP_Proximity &proximity)
     WriteBlock(&pkt_proximity, sizeof(pkt_proximity));
 }
+
+void DataFlash_Class::Log_Write_SRTL(uint8_t active, uint16_t num_points, uint16_t max_points, uint8_t action, const Vector3f& breadcrumb)
```
ok, active made bool in follow up commit.

rmackay9 commented on libraries/DataFlash/LogFile.cpp:
```diff
@@ -1967,3 +1967,19 @@ void DataFlash_Class::Log_Write_Proximity(AP_Proximity &proximity)
+        num_points      : num_points,
+        max_points      : max_points,
+        action          : action,
+        N               : breadcrumb.x,
```
As discussed, i'll ignore this one 'cuz I can't change the fields of a Vector3f.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+/*
```
OK, renamed to AP_SafeRTL in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+extern const AP_HAL::HAL& hal;
+
+const AP_Param::GroupInfo AC_SafeRTL::var_info[] = {
+    // @Param: ACCURACY
```
Not sure I agree with renaming from ACCURACY to INTERVAL.  For me "interval" is normally a time thing - and the interval of saving points is actually 3hz at the moment (controlled from the vehicle code actually).  "accuracy" is a better word I think because it's the minimum distance between points and I think it may also be used as a minimum distance during simplification.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+const AP_Param::GroupInfo AC_SafeRTL::var_info[] = {
+    // @Param: ACCURACY
+    // @DisplayName: SafeRTL _accuracy
+    // @Description: SafeRTL _accuracy. The distance flown (in meters) before a new point is dropped.
```
Added Units and fixed description in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    // @User: Advanced
+    AP_GROUPINFO("ACCURACY", 0, AC_SafeRTL, _accuracy, SAFERTL_ACCURACY_DEFAULT),
+
+    // @Param: PATH_LEN
```
I agree.  I've renamed the parameter to _POINTS_MAX in a follow up commit.  By the way, I prefer to always have the "_MAX" (or _MIN or units) at the end of the parameter name.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+
+    // @Param: PATH_LEN
+    // @DisplayName: SafeRTL path length
+    // @Description: SafeRTL path length. The maximum number of points saved by SafeRTL. Set to 0 to disable SafeRTL.
```
Sure, added, "100 points consumed 3k of memory" to parameter description in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+*    decides how long they will run before saving their state and returning. They
+*    are both anytime algorithms, which is helpful when memory is filling up and
+*    we just need to quickly identify a handful of points which can be deleted.
+*    The algorithms can also report if they are done. If not, the thorough cleanup
```
i've tried to improve the wording in a follow up commit.  I met let @squilter to try and improve it further.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+
+AC_SafeRTL::AC_SafeRTL(const AP_AHRS& ahrs) :
+    _ahrs(ahrs),
+    _active(false),
```
fixed in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+AC_SafeRTL::AC_SafeRTL(const AP_AHRS& ahrs) :
+    _ahrs(ahrs),
+    _active(false),
+    _simplification_stack_last_index(-1),
```
member init moved to .cpp in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    _prunable_loops_last_index(-1)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _simplification_bitmask = std::bitset<SAFERTL_PATH_LEN_MAX>().set(); //initialize to 0b1111...
```
changed to use Bitmask in follow up commit

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _simplification_bitmask = std::bitset<SAFERTL_PATH_LEN_MAX>().set(); //initialize to 0b1111...
+    _time_of_last_good_position = AP_HAL::millis();
```
fixed in follow up commit

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+// initialise safe rtl including setting up background processes
+void AC_SafeRTL::init()
+{
+    // check if user has disabled SafeRTL
```
SafeRTL defaults to on so I'll leave these comments are they are.  Note that while SafeRTL is enabled, it's not used in place of regular RTL for now.  It's a separate flight mode.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+{
+    // check if user has disabled SafeRTL
+    if (_desired_path_len == 0 || is_zero(_accuracy)) {
+        _active = false;
```
fixed in follow up commit

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        return;
+    }
+
+    // protect against repeated call to init
```
removed "_initialised = false" in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    _desired_path_len = MIN(SAFERTL_PATH_LEN_MAX, _desired_path_len);
```
ok, added constraint in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    _desired_path_len = MIN(SAFERTL_PATH_LEN_MAX, _desired_path_len);
+
+    // allocate arrays
+    _path = (Vector3f*)malloc(_desired_path_len * sizeof(Vector3f));
```
replaced malloc with calloc in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+
+    // allocate arrays
+    _path = (Vector3f*)malloc(_desired_path_len * sizeof(Vector3f));
+    _prunable_loops = (loop*)malloc(_desired_path_len * sizeof(loop));
```
ok, i've renamed:

- "loop" to "prune_loop_t" 
- "start_finish" to "simplify_start_finish_t"

And then gone another step to shorten variables names by renaming:

- "_simplification_" to "_simplify_"
- "_pruning_" to "_prune_"

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    // allocate arrays
+    _path = (Vector3f*)malloc(_desired_path_len * sizeof(Vector3f));
+    _prunable_loops = (loop*)malloc(_desired_path_len * sizeof(loop));
+    _simplification_stack = (start_finish*)malloc(_desired_path_len * sizeof(start_finish));
```
yes, renamed start-finish to "simplfy_start_finish_t" in follow up commit.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        free(_path);
+        free(_prunable_loops);
+        free(_simplification_stack);
+        _initialised = false;
```
fixed, thanks.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        free(_simplification_stack);
+        _initialised = false;
+        return;
+    } else {
```
removed unnecessary else

peterbarker commented on ArduCopter/control_safe_rtl.cpp:
```diff
@@ -0,0 +1,136 @@
+        if (!last_point) {
+            // go to next point along path
+            wp_nav->set_wp_destination_NED(next_point);
+        } else {   // if this is the last point, we should prepare to land
```
I think we should actually hit that last point rather than short-cutting here.

rmackay9 replied:
i'm a little torn on what to do about the last point.  It's likely to be on the ground so flying straight towards could easily mean we hit the ground before reaching it and then flip over.

squilter replied:
I agree that it's a bit odd, but Randy explains perfectly why I chose to do it that way. It's not that we are ignoring the last point, it's just that we are going to the location exactly 2m above the last point, then landing.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        _initialised = false;
+        return;
+    } else {
+        _current_path_len = _desired_path_len;
```
renamed "_current_path_len" to "_path_points_max".

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        _current_path_len = _desired_path_len;
+        _initialised = true;
+    }
+
```
I might skip telling the user.  I see dataflash does that but what if every subsystem did that.  Most users won't know what to do with the info.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+}
+
+// clear return path and set home location.  This should be called as part of the arming procedure
+void AC_SafeRTL::reset_path(bool position_ok)
```
yes, it's important we have the position, so fixed comments.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    reset_path(position_ok, current_pos);
+}
+
+void AC_SafeRTL::reset_path(bool position_ok, const Vector3f& current_pos)
```
the reset-path which provides a position is just for the example sketch to use during some canned tests.    It's important the vehicle code to call reset at arming/takeoff.  This essentially tells the SafeRTL library where the return-to location is or disables SafeRTL.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+
+void AC_SafeRTL::reset_path(bool position_ok, const Vector3f& current_pos)
+{
+    if (!_initialised) {
```
I think I've completed an audit of the interface and it seems OK to me.  most functions check "_active" is true and "_active" is only ever true if "_initialised" is also true.  This is because "_reset_path" is the only part of the code that sets "_active" to either _true or _false and this method can only be run if "_initialised" has become true.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    _desired_path_len = MIN(SAFERTL_PATH_LEN_MAX, _desired_path_len);
```
removed unnecessary check, thanks!

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    _desired_path_len = MIN(SAFERTL_PATH_LEN_MAX, _desired_path_len);
+
+    _last_index = 0;
+    _path[_last_index] = current_pos;
```
Considered!  for now I won't do this because we'd need to set "_active = true" before calling update and then I have this tiny worry that one of the background tasks would jump in and start doing stuff.  it would probably be fine but I'm going to leave it as is.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    if (!routine_cleanup()) {
+        _active = false;
+        log_action(SRTL_DEACTIVATED_CLEANUP_FAILED);
+        gcs().send_text(MAV_SEVERITY_WARNING,"SafeRTL: unavailable path cleanup failed");
```
I've updated all the strings to be of the form, "SafeRTL deactivated: path cleanup failed" which I hope is close enough.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    if (position_ok) {
+        // append the new point, if appropriate
```
fixed

WickedShell commented on ArduCopter/control_safe_rtl.cpp:
```diff
@@ -0,0 +1,136 @@
+        // if we're within 4m of the landing point, just skip right to pre-land-positioning
+        // otherwise, begin a normal safertl procedure from the start
+        if (HYPOT(current_pos, g2.safe_rtl.get_point(0)) <= 4.0f) {
+            current_pos[2] -= 2.0f; // go to the point two meters above the landing spot, then start landing.
```
As far as I can tell this is a noop? `current_pos` doesn't looked to be used anywhere after this is set.

rmackay9 replied:
you're right.  I'm sure this is my mistake because I added the stopping-point stuff.  I think I'm just going to remove all this and let it do a regular SafeRTL even if it's within 4m of the target location.

WickedShell commented on libraries/AP_Common/Bitmask.h:
```diff
@@ -65,7 +72,16 @@ class Bitmask {
+    uint16_t count() {
+        uint16_t sum = 0;
+        for (uint16_t i=0; i<numwords; i++) {
+            sum += bits[i];
```
Am I missing something? I don't see this counting the number of bits set, rather it just appears to be adding all the words up, which is not equal to the number of bits. (IE 0x2 + 0x2 will give you 0x4, even though only two bits are set).

rmackay9 replied:
You're right.  I've changed this to use the __builtin_popcount feature.  I have serious concerns though that there will be a problem because Bitmask may allocate slightly more bits that the user has asked for (i.e. if you ask for 33 you get 64).  This means the ::count can return a number higher than the total number of bits the user requested and inside AP_SafeRTL::routine_cleanup this leads to an error.

squilter replied:
The bitmask can safely be longer than the path, as long as the default for the bitmask is 1111...

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+
+            log_action(SRTL_POINT_ADD, current_pos);
+
+            // if cleanup algorithms are finished (and therefore not running), reset them
```
yes, that makes it easier to understand.  I've split the functions in a follow up commit.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,580 @@
+}
+
+/**
+*  Run this method only when preparing to initiate the RTL procedure. Returns a
```
Comment is incorrect, this does not return a pointer.

rmackay9 replied:
Yes, Peter found this too.  fixed.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+    _points_max = constrain_int16(_points_max, 0, SAFERTL_POINTS_MAX);
+
+    // check if user has disabled SafeRTL
+    if (_points_max == 0 || is_zero(_accuracy)) {
```
You should be checking that `!is_positive(accuracy)` instead of zero. (Just to rule out any negative accuracy numbers that might happen by accident).

rmackay9 replied:
great, thanks.  fixed.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+{
+    // this should never happen but just in case
+    if (!_active) {
+        return false;
```
This is problematic in that the error condition is indistinguishable from the normal return value of false. If this is ever hit it the caller will think they have a valid location to go to, but it will just be garbage on the stack. At a minimum if this check is needed then it should set the point to be the current location and the vehicle will at least be stuck there.

rmackay9 replied:
this is fixed in a reworked pop_point.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+
+/**
+*    Simplifies a 3D path, according to the Ramer-Douglas-Peucker algorithm.
+*    Returns the number of items which were removed. end_index is the index of the last element in the path.
```
The return is void, the comment is incorrect.

rmackay9 replied:
corrected comment, thanks.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+
+    // allocate arrays
+    _path = (Vector3f*)calloc(_points_max, sizeof(Vector3f));
+    _prunable_loops = (prune_loop_t*)calloc(_points_max, sizeof(prune_loop_t));
```
Possible memory savings: I don't see why we need points_max for either the simplify stack or prunable loops, (I could easily be missing something). However at a minimum isn't a loop by definition 3 points, so you could have at most _points_max/3 (assuming no redundant encoding of a starting index as being a loop to multiple end points). Similarly for the simplification stack. Given that it's a background process to do the simplification/pruning I would imagine we could keep both of these quite small (say 10, which is the normal amount pruned/simplified in a pass anyways) to maximize the number of points in the path, and just trust the fact that we will be able to do the processing later?

squilter replied:
fixed

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+{
+    // We only do a routine cleanup if the memory is almost full. Cleanup deletes
+    // points which are potentially useful, so it would be bad to clean up if we don't have to
+    if (_last_index < _path_points_max - 10) {
```
I'm not sure we really want to defer this till the buffers almost full. It has the effect of causing the system to have to spend more time (if it simplifies the first points/anything early in the loop), and increases the amount of time it takes to make it through the simplification/cleanup algorithms. I'd suggest we should always take a stab at simplifying/delooping if there are say 20 new points since we last tried?

rmackay9 replied:
I think the "routine_cleanup" (and similarly "thorough_cleanup") actually remove points from the path.  The hardwork of identifying the affected points is done by the background "detect_loops" and/or "detect_simplifications".

I think these "detect_" methods run whenever there are more than 2 or 3 points.
I agree that waiting until we only have 10 points left in the buffer seems like we're waiting until too late.  10 points could very easily be filled in just 3 seconds.

I agree that at least identifying the loops and potential simplifications makes sense.

squilter replied:
> I agree that waiting until we only have 10 points left in the buffer seems like we're waiting until too late.

The longer we wait, the more likely that the backgrouund detect_ methods have found something to simplify or prune.

> 10 points could very easily be filled in just 3 seconds.

That depends on the `SRTL_ACCURACY` parameter, but at the default of 2, we'd have to be travelling at 7m/s for that to happen. But that's not really a problem because the `routine_cleanup` would delete many points in a matter of microseconds.

WickedShell replied:
> The longer we wait, the more likely that the backgrouund detect_ methods have found something to simplify or prune.
The point is to run the simplification algorithm if we have enough stuff that simplifying would actually result in a useful piece of work being done.

Hypothetically doesn't this also mean if you only generate say 180 points but almost all of them are loops, won't the copter currently fly them all backwards rather then taking the simplified/looped path back (effectively increasing the flight time to do the RTL unnecessarily?)

rmackay9 replied:
@WickedShell, no, there's a "thorough_cleanup" that happens as the first step of RTL so the path will definitely be simplified and pruned before the return journey begins.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+*   Otherwise, it will run a cleanup, based on info computed by the background methods, detect_simplifications() and detect_loops().
+*   If no cleanup is possible, this method returns false. This should be treated as an error condition.
+*/
+bool AP_SafeRTL::routine_cleanup()
```
I must be missing something, but I can't find it so I'm going to ask anyways.

As far as I can tell calling this unless we have finished simplifying/delooping is unsafe as the IO thread may be partially through it's loop, and then this will change the underlying structure it's peaking at. I don't see any semaphores/locks on this so it looks like a race condition to me.

rmackay9 replied:
Yes, I think you're right.  @squilter.

squilter replied:
`routine_cleanup` reads the result of the detect_* algorithms, deletes some points from the loop, and then resets the result of the detect_algorithms.

Yes, it is possible that `routine_cleanup` starts deleting points from the path at the same time that detect_* is reading those points. But there are enough checks in there to prevent any division-by-zero or other problems. The worst that can happen when `detect_loops` or `detect_simplifications` checks a malformed path is that it returns an incorrect result (no system lock-up). But the result is deleted immediately after anyways, so there's no way that we would ever read this bad result.

WickedShell replied:
I'm not at all convinced that it's even remotely safe to ignore threading issues here, I don't have time at the moment to trace through it to give you a concrete example, but from the top level when we shift points around while we are in the midst of detecting a loop (say because we simplified) and we did detect a loop but the underlying points shifted we would then emit this as a detected loop and zero out the wrong indexs.

I haven't proven it conclusively, but I think the code is much to complex to be able to wave that any incorrect result couldn't propagate further, and even if that's true now it will be a massive burden to maintain that over time. I think it's definitely worth providing proper synchronization over this.

squilter replied:
Okay, you've convinced me. We need some locks in there.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+/**
+*  Returns the closest distance from a point to a 3D line. The line is defined by any 2 points
+*/
+float AP_SafeRTL::point_line_dist(const Vector3f &point, const Vector3f &line1, const Vector3f &line2)
```
This looks like it might be a decent candidate to move into AP_Math so that it can be used elsewhere as needed.

squilter replied:
agreed

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+    int16_t i = 0;
+    int16_t j = 0;
+    int16_t removed = 0;
+    while (++i <= _last_index) { // never removes the first item. This should always be {0,0,0}
```
It's elsewhere implied that the first item isn't always {0, 0, 0} ArduCopter/control_safe_rtl.cpp:19 is it always going to be {0, 0, 0}? If so get_point() should be moved to a private helper, and control_safe_rtl should just compare against the length of the point it's looking at.

rmackay9 replied:
That comment is incorrect so I've removed it.  the first point will be the vehicle's offset from the ekf origin at the time it arms.  That will most definitely not be 0,0,0 although it may be close to it.

WickedShell commented on libraries/DataFlash/LogStructure.h:
```diff
@@ -873,6 +873,18 @@ struct PACKED log_Proximity {
+    LOG_PACKET_HEADER;
+    uint64_t time_us;
+    uint8_t active;
+    uint16_t num_points;
```
These aren't uint16_t's elsewhere they are int16_t and are being type converted for logging only. They should be consistent (and given that it's used for a counter it's probably better to change the definition in SafeRTL to be uint16_t).

rmackay9 replied:
Yes, Peter has brought  up as well that we should be using uint16_t for all our indexing into the arrays instead of int16_t.  I plan to change this hopefully tomorrow.  It's quite a fundamental change though so I suspect I will break things so I want to do it last.

squilter replied:
At first I had `uint`s everywhere. But then I decided to change `_simplify_stack_last_index`, `_last_index` and `_prunable_loops_last_index`, in order to allow them to store a value of -1 (meaning that there was is not a single thing in the arrays they represent). But then the compiler got angry at me for lots of comparisons between `uint`s and `int`s, so I changed pretty much everything.

So maybe the best fix is to change everything to a `uint` and find some better way to represent that an array is empty.

rmackay9 replied:
@squilter, yes, i think we should change all the indexes back to uint16_t.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.h:
```diff
@@ -0,0 +1,152 @@
+// definitions and macros
+#define SAFERTL_ACCURACY_DEFAULT        2.0f    // default _ACCURACY parameter value.  Points will be no closer than this distance together.
+#define SAFERTL_POINTS_DEFAULT          150     // default _POINTS parameter value.  High numbers improve path pruning but use more memory and CPU for cleanup. Memory used will be 20bytes * this number.
+#define SAFERTL_POINTS_MAX              500     // the absolute maximum number of points this library can support.
```
It's worth adding a static assert to check that SAFERTL_POINTS_MAX is < int16_t max. (Or uint16_t if it gets swapped) otherwise someone on a linux board could easily raise this then be surprised when it doesn't work/they have far less then they expected to.

rmackay9 replied:
My guess is the CPU load will stop us from increasing the number of points long before we get to int16_t max.  Maybe I'm wrong.. we haven't really investigated the memory and CPU limits and do they get worse linearly or exponentially?  Note to @squilter!

squilter replied:
The slowest algorithm is `detect_loops`, which runs in O(n^2) time. On a pixhawk, I think there's no way that it could store a path longer than a couple hundred points. I never did any testing with extremely large paths on a good cpu, but my hunch is that CPU load would be too great to allow a path length anywhere close to 16,000.

khancyr commented on libraries/AP_SafeRTL/AP_SafeRTL.h:
this should go in ap_math too, doesn't it ?

khancyr commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+        area_squared = 0.0f;
+    }
+    float area = safe_sqrt(area_squared);
+    return 2.0f*area/b;
```
need to handle the case where b is 0 ... it shouldn't happend but with float precision and weird people it could !

rmackay9 replied:
Yes, 100% agree.   I've added a is_zero(b) check and return 0.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+
+/**
+*  Run this method only when preparing to initiate the RTL procedure. Returns a
+*  pointer to the cleaned-up path. Returns nullptr if the cleanup algorithms aren't ready yet.
```
fixed comments for thorough_cleanup.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+*  pointer to the cleaned-up path. Returns nullptr if the cleanup algorithms aren't ready yet.
+*  If this happens, just run this method again a bit later.
+*
+*  Probably best not to run this unless cleanup_ready() is returning true
```
I think this is an out of date comment.  It has the checks which confirm the background cleanup tasks have completed.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+/**
+*   Returns true if the list is empty after popping this point.
+*/
+bool AC_SafeRTL::pop_point(Vector3f& point)
```
ok, split pop_point into pop_point and get_num_points.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+*/
+bool AC_SafeRTL::pop_point(Vector3f& point)
+{
+    // this should never happen but just in case
```
I think we've t the checks we need in the new version of pop_point.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+{
+    // this should never happen but just in case
+    if (!_active) {
+        return false;
```
fixed in new version of pop_point.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+{
+    if (!_active || _simplification_complete || _last_index < 2) {
+        return;
+    } else if (_simplification_stack_last_index == -1) {  // if not complete but also nothing to do, we must be restarting.
```
removed unnecessary else.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        // reset to beginning state
+        _simplification_stack[++_simplification_stack_last_index] = start_finish {0, _last_index};
+    }
+    uint32_t start_time = AP_HAL::micros();
```
const the start_time in two places

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        _simplification_stack[++_simplification_stack_last_index] = start_finish {0, _last_index};
+    }
+    uint32_t start_time = AP_HAL::micros();
+    int16_t start_index, end_index;
```
moved start_index and end_index to smaller scope and made const.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+            return;
+        }
+
+        start_finish tmp {}; // initialize to zero to suppress warnings
```
compressed tmp init into one line and added const

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+
+        start_finish tmp {}; // initialize to zero to suppress warnings
+        tmp = _simplification_stack[_simplification_stack_last_index--];
+        start_index = tmp.start;
```
instead i've changed the check so it uses end-index.  it's true that it wastes one assign but it's possible the assignment doesn't even happen after it's all compiled.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        int16_t index = start_index;
+        for (int16_t i = index + 1; i < end_index; i++) {
+            if (_simplification_bitmask[i]) {
+                float dist = point_line_dist(_path[i], _path[start_index], _path[end_index]);
```
dist has been const-ed.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    if (!_active || _pruning_complete || _last_index < 3) {
+        return;
+    }
+    uint32_t start_time = AP_HAL::micros();
```
start-time made const

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+                    return;
+                }
+                // int promotion rules disallow using i+1 and j+1
+                _prunable_loops[++_prunable_loops_last_index] = {(++_pruning_current_i)--,(++j)--,dp.point};
```
@squilter should take on this change.  That line scares me too much to touch.

squilter replied:
The line is supposed to look like this: 
`_prunable_loops[++_prunable_loops_last_index] = {_pruning_current_i + 1, j + 1, dp.point};`.

The problem is that apparently adding 1 to a `uint8_t` will promote the `uint8_t` to a `uint_32_t` before adding one. Casting back down gives annoying compiler errors.

rmackay9 replied:
@squilter, it's a very scary line, I really thing we should split that one up.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+{
+    // We only do a routine cleanup if the memory is almost full. Cleanup deletes
+    // points which are potentially useful, so it would be bad to clean up if we don't have to
+    if (_last_index < _current_path_len - 10) {
```
added MAX() and defintions for SAFERTL_CLEANUP_START_MARGIN and SAFERTL_CLEANUP_POINT_MIN

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    for (int16_t i = 0; i <= _last_index; i++) {
+        if (!_simplification_bitmask[i]) {
+            _simplification_clean_until = MIN(_simplification_clean_until, i-1);
+            if (_path[i] != Vector3f(0.0f, 0.0f, 0.0f)) {
```
Vector3f::is_zero exists apparently so follow up commit uses it instead of this existing test.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+/**
+*   Only prunes loops until $points_to_delete points have been removed. It does not necessarily prune all loops.
+*/
+void AC_SafeRTL::zero_points_by_loops(int16_t points_to_delete)
```
changed arg to uint16_t

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+        _pruning_clean_until = MIN(_pruning_clean_until, l.start_index-1);
+        for (int16_t j = l.start_index; j < l.end_index; j++) {
+            // zero this point if it wasn't already zeroed
+            if (_path[j] != Vector3f(0.0f, 0.0f, 0.0f)) {
```
replaced with is_zero.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+*/
+void AC_SafeRTL::remove_empty_points()
+{
+    int16_t i = 0;
```
renamed i and j to src and dest respectively

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+{
+    int16_t i = 0;
+    int16_t j = 0;
+    int16_t removed = 0;
```
changed "removed" to unsigned.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    int16_t j = 0;
+    int16_t removed = 0;
+    while (++i <= _last_index) { // never removes the first item. This should always be {0,0,0}
+        if (_path[i] != Vector3f(0.0f, 0.0f, 0.0f)) {
```
replaced with is_zero.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.cpp:
```diff
@@ -0,0 +1,573 @@
+    Vector3f line2 = p4-p3;
+    Vector3f line_start_diff = p1-p3; // from the beginning of the second line to the beginning of the first line
+
+    // these don't really have a physical representation. They're only here to break up the longer formulas below.
```
I'll leave this one for @squilter

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.h:
```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <AP_Buffer/AP_Buffer.h>
```
removed AP_Buffer include

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.h:
```diff
@@ -0,0 +1,146 @@
+#include <GCS_MAVLink/GCS.h>
+
+#include <bitset>
+#include <vector>
```
removed bitset and vector includes

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.h:
```diff
@@ -0,0 +1,146 @@
+#include <vector>
+
+// definitions and macros
+#define SAFERTL_ACCURACY_DEFAULT        2.0f    // default _ACCURACY parameter value.  Points will be no closer than this distance together.
```
added units to comments.  normally when we use SI units we don't specify the units in the name.

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.h:
I have not tested the time taken so I leave this to @squilter

rmackay9 commented on libraries/AC_SafeRTL/AC_SafeRTL.h:
i've rename "point" and "halfway_point" to "midpoint".

rmackay9 commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,584 @@
+
+    // allocate arrays
+    _path = (Vector3f*)calloc(_points_max, sizeof(Vector3f));
+    _prunable_loops = (prune_loop_t*)calloc(_points_max, sizeof(prune_loop_t));
```
Yes, i've though the same thing.  I think @squilter should tackle this one.

rmackay9 commented on libraries/AP_SafeRTL/AP_SafeRTL.h:
yes, I plan to move it to AP_Math

rmackay9 commented:
Thanks very much for the review.  These are the things I will focus on probably tomorrow and try and resolve (at least the first two):

- need semaphore to protect zeroing/removing points while detect is running
- indexes should be uint16_t
- move HYPOT, segment_segment_dist, point_line_dist to AP_Math
- can we optimise the memory? _prunable_loops, _simplify_stack are likely too big
- can we optimise CPU?  HYPOT, segment_segment_dist, point_line_dist
- need to determine memory and CPU limits as we raise max number of points

squilter commented on libraries/AC_SafeRTL/AC_SafeRTL.h:
Both `detect_simplifications` and `detect_loops` will only allow themselves to be interrupted at place where it makes sense to pause computation. This means that sometimes `detect_simplification` might run for 1ms or more before pausing computation, even though `SAFERTL_SIMPLIFICATION_TIME_US ==  200`. So these numbers frankly aren't very useful. All it really means is that if `detect_simplifications` is called, it won't return for at least 200 us (unless, of course, there is nothing for it to do).

The 'optimal' number would probably depend on the path length, since one runs in n*log(n) time and the other in n^2 time. But as long as the timing is not super critical (read: it isn't running in the main loop), it doesn't seem to matter much.

squilter commented:
> - need semaphore to protect zeroing/removing points while detect is running

I'm not sure this is necessary. See my comment above.

> - indexes should be uint16_t

The complication is that some indexes store a -1 to mean that the array is empty. You also shouldn't mix and match `uint`s an `int`s because the compiler will produce a lot of warnings. So you'll have to turn pretty much all `int`s to `uint`s and then find some other way to represent that the arrays are empty.

> - move HYPOT, segment_segment_dist, point_line_dist to AP_Math

Agreed.

> - can we optimise the memory? _prunable_loops, _simplify_stack are likely too big

`_prunable_loops` is where the `detect_loops` algorithm writes down the loops it has detected. The way it's currently written, `_prunable_loops` is long enough to allow for detection of n/4 loops where n is the length of the `_path`. We could easily make this array smaller (like n/8), but that might not be enough space. (Don't worry - there's checks to prevent it from writing out of bounds)

`_simplify_stack` is actually used by `_detect_simplifications` while it is running the algorithm. The algorithm would not be able to run properly if it is not guaranteed enough space to write down everything it needs to write down.

> - can we optimise CPU? HYPOT, segment_segment_dist, point_line_dist

Maybe. The most expensive method (by far) is `_detect_loops`, so that would be a good place to look first. That one makes a lot of calls to `segment_segment_dist`.

rmackay9 commented:
@squilter re _prunable loops, it's calloc'd to have a prune_loop_t for each point.  So it's sized at "n" rather than "n/4".  Same goes for _simplify_stack.

By the way, I think all the changes have broken it somehow so I'll look into that now.  Hopefully it's some small bug because most of the changes I did yesterday in response to the review have been non-functional changes.

squilter commented:
Oops, that's not right. It looks like I introduced that mistake myself before you took over. This:
```
_prunable_loops = (loop*) calloc(_current_path_len * sizeof(loop));
_simplification_stack = (start_finish*) calloc(_current_path_len * sizeof(start_finish));
```
Should be changed to this:
```
should be changed to this
_prunable_loops = (loop*) calloc(_current_path_len * sizeof(loop) * SAFERTL_LOOP_BUFFER_LEN_MULT);
_simplification_stack = (start_finish*) calloc(_current_path_len * sizeof(start_finish) * SAFERTL_SIMPLIFICATION_STACK_LEN_MULT);
```

rmackay9 commented:
@squilter, Ok great.  I'll add that mult back in there.

I think we should break up the remaining efforts to get this in.  I'm over in the GSoC gitter channel if you're around.
In the very short term, I'm going to figure out how I've broken it.  I've found one problem but there's another - the pruning isn't working anymore.

rmackay9 commented:
@squilter, as promised here's the hand-over from my side:

Completed the following:

- reduced memory consumption by reducing _prunable_loops, _simplify_stack (as discussed with you)
- indexes should be uint16_t
- simplified this difficult line, "_prunable_loops[++_prunable_loops_last_index] = {(++_pruning_current_i)--,(++j)--,dp.point};"

These items are still outstanding:

- need semaphore to protect zeroing/removing points while detect is running
- move HYPOT, segment_segment_dist, point_line_dist to AP_Math
- can we optimise CPU? HYPOT, segment_segment_dist, point_line_dist
- need to determine memory and CPU limits as we raise max number of points


Here's the latest work: https://github.com/rmackay9/rmackay9-ardupilot/tree/squilter-safe-rtl5
Feel free to raise a new PR if you like or we can wait until we've gotten through all remaining issues.

Re the semaphore issue, I think this might work:

- combine detect_simplifications and detect_loops to run in a single thread
- run routine_cleanup and thorough_cleanup (i..e the zero-points-xxx, remove-empty-point, etc) in the background thread instead of from update.
- add an "add_point" method.  this will be called from update() and also from the bottom of reset_path() and, as the name suggests, it simply adds a point to the end of _path.
- create a new semaphore (hal.util->new_semaphore()).  Have a look at the AP_Airspeed library for how it can be used.  This will be used by any part of the code that wants to modify the _path_points_count which I think means just remove_empty_points and the new add_point method.  I think the new add_point should sem->take(<short-period-of-time>) and if it can't get it it should log an error and return.  The remove-empty-point could block for longer and if it fails.. i'm not sure what it should do.  it shouldn't really fail..

squilter commented:
@rmackay9 I just picked up squilter-safertl-5, and there seems to be a problem. Then I reverted to -4 and it had the same problem. The problem is that once I switch to SafeRTL mode, it prematurely leaves `SafeRTL_PathFollow` to enter `SafeRTL_PreLandPosition`. I suspect that many points are being incorrectly deleted. I'll look into that now, probably best to fix that before making those other changes. I'm also trying to fix the example sketch.

squilter commented:
Okay that's resolved, just a single line change in `Bitmask.h`.

The example sketch is fixed as well. That one was slightly more complicated. Take a look, because you can probably think of a cleaner way to do that.

The changes are at `squilter/squilter-safe-rtl5`. Here's a diff:

https://github.com/rmackay9/rmackay9-ardupilot/compare/squilter-safe-rtl5...squilter:squilter-safe-rtl5

squilter commented:
It's safe to append points while the detect_* algorithms are running. The danger is when zeroing or deleting existing points. Does that detail make the semaphore thing easier?

rmackay9 commented:
@squilter, that's great stuff.  I've pulled it into my squilter-safe-rtl5 branch.

Re the semaphore thing, yes, the details you've given make it easier.  I'm happy to take this on adding the semaphore.  I didn't spend much time on it today but I suspect I'll get the bulk of the semaphore based protection added tomorrow.

rmackay9 commented:
@squilter,

I think I've solved @WickedShell's request that we protect against the vehicle code and cleanup threads from trying to access some of the same variables.  The modified branch is here: https://github.com/rmackay9/rmackay9-ardupilot/tree/squilter-safe-rtl7

It was surprisingly difficult. I did the following:

- run the detect-simplify and detect-loop algorithms from a single thread instead of two separate threads.  These are called from a new run_background_cleanup method.
- added a semaphore which is used before any code tries to access or modify the _path_points_count variable.  This includes the add-point and pop-point methods which are called from the vehicle code and also all the functions that run as part of "run_background_cleanup".
- we reduce the requirements to use the semaphore by keeping local copies of the path-points-count for the simplify and loop detection algorithms (i.e _simplify_path_points_count, _prune_path_points_count).

I also did the following:

- split up the detect-loops so that it stores the "j" value between iterations and thus is more granular and is less likely to run too long.
- fixed a bug in the prune algorithm's j index.  I think it could sometimes overrun the array.

I haven't tested this thoroughly although from a quick test in the simulator it still seems to work.  I haven't checked the example sketch at all.

Maybe you could have a look and if you think it's OK we can re-raise the PR and ask @OXINARF, @peterbarker and @WickedShell to re-test.

squilter commented:
Nice job! Here's some quick thoughts:

> - run the detect-simplify and detect-loop algorithms from a single thread instead of two separate threads. These are called from a new run_background_cleanup method.

I had it like that originally, but then split them up. The reason is that when they are in the same method, the scheduler cannot possibly schedule any io operations in between them, it has to wait until both methods have run.

> - split up the detect-loops so that it stores the "j" value between iterations and thus is more granular and is less likely to run too long.

Good idea.

> - removed _simplify_clean_until and _prune_clean_until because I think they could never actually reduce calculations required because any time a point was removed, a reset-simplification or reset-prune was called. Maybe I was misunderstanding something though.

Oh yeah. Originally, I think it kinda worked. But then while debugging one day I decided to effectively disable it by resetting it to zero on reset. It's probably best to leave it out for now.

squilter commented:
The example compiles, but does not work. You changed the interface enough that the example sketch needs to be changed. I'm not sure how easy it would be to run pruning and simplification separately now. Can you take a look?

Did you still want to move `HYPOT`, `point_line_dist` and `segment_segment_dist` to AP_Math?

After that I think it would be a good idea to start over with a new PR.

rmackay9 commented:
Closing this in favour of the new PR. https://github.com/ArduPilot/ardupilot/pull/6868

rmackay9 commented:
@squilter, I've fixed up the example sketch so it works now.  It produces an error on the prune step but I think the issue is the "test_path_after_pruning" is incorrect.  I could be wrong but while making the prune-loops more granular I found and fixed an indexing error (the "j" could go beyond the end of the path) so my hope is that fixing this bug solved the problem.

Maybe you can have a look at the new PR and see if you see any issues.  I think once we're sure it's "safe" we should look into optimisations.  Now I know you don't have much time left with us this summer, but I think the distance-to-segment and segment_segment_dist should be looked over to potential optimisations.  Another area to optimise if figuring out a way to keep the loop info when we remove points from the path.
